### DISCUSSION TITLE
How do you know when a source is done loading from a url?

### DISCUSSION THREAD

**Discussion by togakangaroo** [uber/react-map-gl#2401] [Q&A]
I am passing a url to my source data to fetch and display a potentially large geojson file. After it is done loading and rendering I want to fit my map to the data.

The issue I'm running into is that no event seems to be reliable. I get multiple `dataloading` and `data` and `sourcedata` events *as I pan and zoom*, not just when I query new data. Yes, I am checking `sourceId` and the loading status and that it is not for metadata. At least some of the events received as I move around the map are the same as some of the events I receive after a data source load are indistinguishable from the events I receive as I pan and move. I know there aren't multiple requests being sent out on the network though.

The thing is that there isn't enough documentation either here or in mapbox to really understand if this is intentional or unexpected behavior. It seems to be so vested in the use case of either passing. in your geojson directly or using tilesets that other options are really left unclear.

**Comment by michaelcgorman** [comment]
_In theory_, you should be able to subscribe to `sourcedata` events and check the `isSourceLoaded` property like so:

```TSX
import { Map, type MapSourceDataEvent } from 'react-map-gl';

const handleSourceData = (e: MapSourceDataEvent) => {
  if (e.isSourceLoaded) {
    // ...
  }
};

const CustomMap = () => {
  return (
    <Map onSourceData={handleSourceData}>
      // ...
    </Map>
  );
};
```

But I say "in theory" because that's not actually true. The problem is that Mapbox doesn't always fire a `sourcedata` event when a source finishes loading.

The only solution I've found is to set a timer/interval and repeatedly ask the map whether the source has finished loading, e.g.:

```TSX
import { Map, type MapSourceDataEvent } from 'react-map-gl';

const MAXIMUM_DELAY_MS = 100;

const sourceTimers: Record<string, number> = {};

const executeSourceDataHandler = (e: MapSourceDataEvent) => {
  // ...
};

const handleSourceData = (e: MapSourceDataEvent, shouldSetTimer = true) => {
  if (e.isSourceLoaded || e.target.isSourceLoaded(e.sourceId)) {
    if (sourceTimers[e.sourceId] !== undefined) {
      window.clearInterval(sourceTimers[e.sourceId]);
      delete sourceTimers[e.sourceId];
    }
    executeSourceDataHandler(e);
  } else if (shouldSetTimer) {
    if (sourceTimers[e.sourceId] !== undefined) {
      window.clearInterval(sourceTimers[e.sourceId]);
    }
    sourceTimers[e.sourceId] = window.setInterval(handleSourceDataEvent, MAXIMUM_DELAY_MS, e, false);
  }
};

const CustomMap = () => {
  return (
    <Map onSourceData={handleSourceData}>
      // ...
    </Map>
  );
};
```

Is it pretty? Absolutely not. But it does reliably execute the handler.
